### PR TITLE
MB-926 introducing global error based flow control and related enhancements.

### PR DESCRIPTION
**Pull request by ramith**
This is the initial changes for MB-926 jira.

### CODE REVIEW

**Comment by ramith** [comment]
Merging this one. I will commit code review changes in a later commit.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesChannel.java [hunk 1]
```diff
@@ -87,22 +87,44 @@
     private boolean flowControlEnabled;
 
     /**
-     * Track global flow control status
+     * Indicates if the buffer based flow controls is enabled a global level.
      */
-    private boolean globalFlowControlEnabled;
+    private boolean globalBufferBasedFlowControlEnabled;
+
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesChannel.java]: Let's describe what lobalBufferBasedFlowControl is as well

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesChannel.java [hunk 2]
```diff
@@ -146,6 +188,20 @@ private synchronized void unblockLocalChannel() {
...
 
     /**
+     * Notify local channel to block channel temporary
+     */
+    private synchronized void blockLocalChannel() {
+        if (!flowControlEnabled) {
+            flowControlEnabled = true;
+            listener.block();
+            scheduledFlowControlTimeoutFuture = executor.schedule(flowControlTimeoutTask, 1, TimeUnit.MINUTES);
+
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesChannel.java]: Better to describe in a comment what we are doing here

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesChannel.java [hunk 3]
```diff
@@ -185,7 +234,9 @@ public void recordRemovalFromBuffer(int size) {
 
         int count = messagesOnBuffer.addAndGet(-size);
 
-        if (flowControlEnabled && !globalFlowControlEnabled && count <= flowControlLowLimit) {
+        if (flowControlEnabled && (!globalBufferBasedFlowControlEnabled) 
+                               && (count <= flowControlLowLimit) 
+                               && (! globalErrorBasedFlowControlEnabled)) {
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesChannel.java]: This condition seems to be used everywhere. Let's make a private method and reuse? Then it is handy if we have to change the logic someday..

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/FlowControlManager.java [hunk 4]
```diff
@@ -163,54 +173,90 @@ public ScheduledExecutorService getScheduledExecutor() {
...
+     *
+     * @param size
+     *         Number of items removed from buffer
      */
-    private synchronized void blockListeners() {
-        if (!globalFlowControlEnabled) {
-            globalFlowControlEnabled = true;
+    public void notifyRemoval(int size) {
+        int count = messagesOnGlobalBuffer.addAndGet(-size);
+
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/FlowControlManager.java]: I did not understand this line

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/inbound/MessageWriter.java [hunk 5]
```diff
@@ -59,32 +81,94 @@ public void onEvent(final List<InboundEventContainer> eventList) throws Exceptio
...
+            // This happens when message store becomes unavailable and came back online after a while.
+            // Previous message insert batch was successful in Database but connection was dropped before
+            // client (= MB) got to know about it ( aka. commit worked).
+            // Now message writer goes and inserts same batch again -> results in failures in batch update.
+            // Therefore here we remove conflicting message parts (which are probably already in the database).
+            //currentMessageList.removeAll(batchInsertEx.getFailedInserts());
+            previouslyFailedMessageList.addAll(currentMessageList);
+            currentMessageList.clear();
+        } catch (Exception ex) {
+            log.warn("unable to store messages, probably due to errors in message stores. messages count : " + currentMessageList.size());
```
Comment (pumudu88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/distruptor/inbound/MessageWriter.java]: can we break this line into two?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSAndesContextStoreImpl.java [hunk 6]
```diff
@@ -913,4 +1031,17 @@ protected void close(ResultSet resultSet, String task) {
             }
         }
     }
+
+    @Override
+    public boolean isOperational(String testString, long testTime) {
```
Comment (pumudu88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSAndesContextStoreImpl.java]: can we add a method description?

**Pull request merged**